### ISSUE TITLE
emoji symbols not supported

### ISSUE THREAD

**Issue by runarberg** [mathjax/MathJax#2369]
Using `mml-chtml` emoji symbols in MathML input won’t translate into typeset chtml.

See screenshot below from https://runarberg.github.io/mathup if you turn the MathJax polyfill on and scroll down to the reference section on identifiers you will see this:

![MathJax removes the emoji symbols](https://user-images.githubusercontent.com/3984600/76725965-0a1b0480-670d-11ea-996c-f13e058fd324.png)

However with native MathML rendering this is what you see:

![Native MathML show emoji](https://user-images.githubusercontent.com/3984600/76726389-36835080-670e-11ea-810a-62c3eefe7313.png)

I suspect this is because typesetting iterates over each UTF-16 char code as opposed to UTF-8 character.

**Comment by runarberg** [comment]
Ironically I just found out that [殺](https://en.wiktionary.org/wiki/%E6%AE%BA) means “to kill” which is a bit unfortunate given it comes from the 🥰 (“smiling face with hearts”) emoji.

**Comment by dpvc** [comment]
Yes, I came across this a week or so ago myself, and have a fix for it as part of a larger font-based pull request that I will be making shortly.

> I suspect this is because typesetting iterates over each UTF-16 char code as opposed to UTF-8 character.

This is the right idea, but not quite it.  Javascript strings are all UT-16 regardless of the original encoding in the HTML page, and MathJax does properly iterate over multi-character UTF-16 characters.  The failure is at the output end of things, not the input.  It doesn't render multi-character UTF-16 properly.  This is because it was using `String.fromCharCode()` rather than `String-fromCodePoint()`, so multi-character code points get treated as single-character ones (dropping the upper character), which is what you are seeing here.

This will be fixed in 3.0.2, but in the meantime, you could use the following configuration

``` html
    <script>
    MathJax = {
      startup: {
        ready() {
          const CHTMLTextNode = MathJax._.output.chtml.Wrappers.TextNode.CHTMLTextNode;
          CHTMLTextNode.prototype.toCHTML = function (parent) {
            this.markUsed();
            const adaptor = this.adaptor;
            const variant = this.parent.variant;
            const text = this.node.getText();
            if (variant === '-explicitFont') {
              const font = this.jax.getFontData(this.parent.styles);
              adaptor.append(parent, this.jax.unknownText(text, variant, font));
            } else {
              const c = this.parent.stretch.c;
              const chars = this.parent.remapChars(c ? [c] : this.unicodeChars(text));
              for (const n of chars) {
                const data = this.getVariantChar(variant, n)[3];
                const node = (data.unknown ?
                              this.jax.unknownText(String.fromCodePoint(n), variant) :
                              this.html('mjx-c', {class: this.char(n)}));
                adaptor.append(parent, node);
                data.used = true;
              }
            }
          }
        }
      }
    }
    </script>
```

to apply the needed patch for the CHTML output (if you want SVG output, there is a similar patch needed for the SVG renderer).
